File: lib/php/ctrl/fm-action.php
<?php
/*
 * This file is part of the iFilemanager package.
 * (c) 2010-2011 Stotskiy Sergiy <serjo@freaksidea.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
if (!defined('SJ_IS_ADMIN')) {
    header('Location: http://www.google.com');
    exit;
}

$path   = trim($_REQUEST['path']);
$action = $_REQUEST['action'];
$base_work_space = $sjConfig['root'];
$path   = $base_work_space . $path;
$path   = rtrim($path, DIRECTORY_SEPARATOR) . DIRECTORY_SEPARATOR;

try {
    if (!pathIn($path, $base_work_space)
        || isset($sjConfig['allowed_actions'])
        && !in_array($action, $sjConfig['allowed_actions'])
    ) {
        throw new sjException($_SYSTEM['i18n']->__('Access denied'));
    }

    $fs = new iFilesystem();
    $fm = iFilemanager::create()
        ->setFilesystem($fs->setI18n($_SYSTEM['i18n']))
        ->setI18n($_SYSTEM['i18n']);

    $files = array();
    $has_files = isset($_REQUEST['files']) && is_array($_REQUEST['files']);

    if ($has_files) {
        $files = $_REQUEST['files']; // name of files

        if (isset($_REQUEST['baseDir'])) {
            $base_dir =  trim($_REQUEST['baseDir']); // in the the path have symbol '/'
            $base_dir = $base_work_space . $base_dir;
            if (!pathIn($base_dir, $base_work_space)) {
                throw new sjException($_SYSTEM['i18n']->__('Access denied'));
            }
        } else {
            $base_dir = $path;
        }

        foreach ($files as &$file) {
            $file = $base_dir . $file;
        }

        $fm->import($files);
    }
    switch($action){
        case 'paste':
            if (!$has_files) {
                throw new sjException($_SYSTEM['i18n']->__('Unable to process request'));
            }

            $fm->paste($path, array(
                'dynamic_name' => true,
                'move'         => isset($_REQUEST['onlyCopy']) && !$_REQUEST['onlyCopy']
            ));
        break;
        case 'remove':
            if (!$has_files) {
                throw new sjException($_SYSTEM['i18n']->__('Unable to process request'));
            }

            $fm->removeAll();
        break;
        case 'download':
            if (!$has_files) {
                $fm->import($path);
            }

            $fm->send('files.zip');
            $_SYSTEM['jsRequest']->reset();
        break;
        case 'perms':
            if (!$has_files) {
                throw new sjException($_SYSTEM['i18n']->__('Unable to process request'));
            }

            if (isset($_REQUEST['fileperms'])) {
                $fileperms = trim($_REQUEST['fileperms']);
            } else {
                $fileperms = false;
                $_SYSTEM['tmpl'] = 'actions/perms';
            }

            if (!$fileperms && $fm->count() == 1) {
                $mode = (int)$fs->getMode(reset($files));
                $_SYSTEM['tmpl_params']['perms_value'] = $mode;
            } elseif (is_numeric($fileperms)) {
                if (strlen($fileperms) < 3) {
                    while (strlen($fileperms) != 3) $fileperms .= '0';
                }
                $fileperms = substr($fileperms, 0, 3);
                $fs->chmod($files, eval("return(0{$fileperms});"));
            }
        break;
        case 'dir_info':
            $filename = $path;
            if ($has_files) {
                $filename = reset($files);
            }
            $stat = $fs->stat($filename);
            $stat['size'] = $fs->formatSize($filename);

            $_SYSTEM['tmpl_params']['property'] = array_slice($stat, 0);
            $_SYSTEM['tmpl'] = 'actions/property';
        break;
        case 'create_dir':
            if (!isset($_REQUEST['dirname']) || !$_REQUEST['dirname']) {
                throw new sjException($_SYSTEM['i18n']->__('Unable to process request'));
            }

            $dirname = trim($_REQUEST['dirname']);

            if (preg_match('/[\/:*?<>|\'"]+/', $dirname)) {
                throw new sjException($_SYSTEM['i18n']->__('File name can not contains following symmbols: \/:*?<>|"\''));
            }

            if (is_dir($path . $dirname)) {
                throw new sjException($_SYSTEM['i18n']->__('Unable to create folder. Folder with this name already exists'));
            }

            if (!$fs->mkdirs($path . $dirname)) {
                throw new sjException($_SYSTEM['i18n']->__('Unable to create folder. Permissions denied'));
            }
        break;
        case 'rename':
            if (!$has_files || !isset($_REQUEST['fileNames']) || empty($_REQUEST['fileNames'])) {
                throw new sjException($_SYSTEM['i18n']->__('Unable to process request'));
            }

            $newFileName = $fs->prepareFilename(reset($_REQUEST['fileNames']));
            $newFileExt  = $fs->getPathInfo($newFileName, PATHINFO_EXTENSION);
            if (!empty($sjConfig['uploader']['allowed_types'])
                && !in_array($newFileExt, $sjConfig['uploader']['allowed_types'])
            ) {
                throw new sjException($_SYSTEM['i18n']->__('Files with extension "%s" does not allowed', $newFileExt));
            }

            $oldFile = reset($files);
            $dirname = dirname($oldFile);
            $fs->rename($oldFile, $dirname . DIRECTORY_SEPARATOR . $newFileName);
        break;
        case 'upload':
            if (empty($_FILES)) {
                throw new sjException($_SYSTEM['i18n']->__('Unable to process request'));
            }

            $fm->import($_FILES)
                ->paste($path, $sjConfig['uploader']);
        break;
    }

    $_RESULT['response']['status'] = 'correct';
    $_RESULT['response']['msg'] = $_SYSTEM['lang']['REQUEST_DONE'];
} catch (sjException $e) {
    $_RESULT['response']['status'] = 'error';
    $_RESULT['response']['msg']    = $e->getMessage();

    // for flash SWFUpload
    if (isset($_REQUEST['print_error']) && $_REQUEST['print_error']) {
        echo $_RESULT['response']['msg'];
    }
}

if (isset($_SYSTEM['tmpl'])) {
    $view = new iView($_SYSTEM['tmpl']);
    $view->render(array_merge(
        $_SYSTEM['tmpl_params'],
        array('lang' => $_SYSTEM['lang'])
    ));
}
?>

File: lib/ruby-rack/ctrl/fm_action.rb
require 'model/file_manager'

module SjFileManager
  class FileController < Controller
    def dispatch
      validate!()

      unless respond_to?(@context.params['action'])
        raise SjException, @context.i18n.__('Unknown action: %s',  @context.params['action'])
      end

      result = send(@context.params['action'])
      if result === false
        return nil
      end

      response = {
        :response => {
          :status => 'correct',
          :msg    => @context.i18n.__('Request was successfully done')
        }
      }

      if result.kind_of?(::Hash)
        result.each do |k, v|
          response[k] = v
        end
      end

      return response
    end

    def paste
      raise SjException, @context.i18n.__('Unable to process request') unless has_files

      have_to_copy = @context.params['onlyCopy']
      manager!.paste(get_path, {
        'dynamic_name' => true,
        'move'         => !have_to_copy || have_to_copy.empty?
      })
    end

    def remove
      fm = manager!
      raise SjException, @context.i18n.__('Unable to process request') if fm.empty?

      fm.remove_all
    end

    def download
      filename = 'files.zip'
      fm = manager!

      if fm.empty?
        fm.import [get_path]
      end

      file = fm.compress(filename)
      @context.response['Content-Type'] = 'application/octet-stream'
      @context.response['Expires'] = Time.now.gmtime.to_s
      @context.response['Content-Disposition'] = 'attachment; filename="%s"' % filename
      @context.response['Accept-Ranges'] = 'bytes'
      @context.response['Content-Length'] = File.size(file.path)

      if @context.request.user_agent.index 'MSIE'
        @context.response['Cache-Control'] = 'must-revalidate, post-check=0, pre-check=0'
        @context.response['Pragma'] = 'public'
      else
        @context.response['Pragma'] = 'no-cache'
      end

      File.open(file.path) do |file|
        @context.response.write file.read
      end
      return false
    end

    def perms
      raise SjException, @context.i18n.__('Unable to process request') unless has_files

      perms = @context.params['fileperms'].to_i

      unless perms == 0
        perms = perms.to_s
        if perms.length < 3
          perms << '0' * (3 - perms.length)
        end

        perms = perms.slice(0, 3).to_i(8);
        manager!.filesystem.chmod(get_files!, perms)
      end
    end

    def dir_info
      fm = manager!
      filename = fm.empty? ? get_path : get_files!.first

      stat = fm.filesystem.stat(filename)
      data = { :mode => nil, :size => nil, :mtime => nil, :type => nil }
      data.each_key do |k|
        data[k] = stat[k]
      end

      data[:size] << @context.i18n.__('b')
      data[:mtime] = @context.i18n.format_date(data[:mtime])
      data[:type].capitalize!

      return { :file_info => data }
    end

    def create_dir
      dirname = @context.params['dirname']
      raise SjException, @context.i18n.__('Unable to process request') if !dirname || dirname.empty?

      raise SjException, @context.i18n.__('File name can not contain following symbols: \/:*?<>|"\'') if dirname =~ /[\/:*?<>|'"]+/

      dirname.strip!
      path = get_path
      puts path
      raise SjException, @context.i18n.__('Unable to create folder. Folder with this name already exists') if File.directory?(path + dirname)

      raise SjException, @context.i18n.__('Unable to create folder. Permissions denied') unless manager!.filesystem.mkdirs(path + dirname)
    end

    def rename
      if !has_files || !@context.params['fileNames'] || @context.params['fileNames'].empty?
          raise SjException, @context.i18n.__('Unable to process request')
      end

      fs = manager!.filesystem

      new_file_name = fs.prepare_filename(@context.params['fileNames'].first[1])
      new_file_ext  = fs.get_pathinfo(new_file_name)[:extension]

      allowed_types = @context.config['uploader']['allowed_types']
      if new_file_ext && !new_file_ext.empty? && allowed_types.kind_of?(::Array) && !allowed_types.include?(new_file_ext)
        raise SjException, @context.i18n.__('Files with extension "%s" does not allowed', new_file_ext)
      end

      old_file = get_files!.first
      dirname  = File.dirname(old_file)
      fs.rename(old_file, dirname + File::Separator + new_file_name)
    end

    def upload
      files = @context.files
      raise SjException, @context.i18n.__('Unable to process request') if !files || files.empty?

      cfg_name = @context.params['use_cfg']
      cfg = @context.config['uploader']

      if cfg['named'] && cfg['named'][cfg_name]
        cfg['images'] = cfg['named'][cfg_name]['images']
        cfg['thumbs'] = cfg['named'][cfg_name]['thumbs']
      end

      fm = manager!(true)
      fm.import files.values
      fm.paste(get_path, cfg)
    end

    private
      def validate!
        unless @context.config['allowed_actions'].nil? || @context.config['allowed_actions'].include?(@context.params['action'])
          raise SjException, @context.i18n.__('Access denied')
        end
      end

      def get_path
        path = @context.config['root'] + get_dir(@context.params['path'])
        path << File::Separator unless path.end_with?(File::Separator)
        return path
      end

      def has_files
        files = get_files_dirty
        files.kind_of?(::Array) && !files.empty?
      end

      def get_files_dirty
        files = @context.params['files']
        if files.kind_of?(::Hash) && !files.empty?
          files = files.values
        end

        return files
      end

      def get_files!
        files = get_files_dirty
        return [] unless has_files

        base_dir = @context.params['baseDir']

        unless base_dir.nil? || base_dir.empty?
          base_dir = get_dir!(base_dir)
        else
          base_dir = get_dir(@context.params['path'])
        end

        base_dir << File::Separator unless base_dir.end_with?(File::Separator)

        base_dir = @context.config['root'] + base_dir
        files.map do |file|
          base_dir + file
        end
      end

      def manager!(empty = false)
        fm = Filemanager.new
        fm.filesystem = Filesystem.new
        fm.filesystem.i18n = @context.i18n

        fm.import get_files! unless empty
        return fm
      end
  end
end
